containers-tests/tests/IntSetValidity.hs

```diff
@@ -24,48 +25,41 @@ nilNeverChildOfBin t =
   case t of
     Nil -> True
     Tip _ _ -> True
-    Bin _ _ l r -> noNilInSet l && noNilInSet r
+    Bin _ l r -> noNilInSet l && noNilInSet r
   where
     noNilInSet t' =
       case t' of
         Nil -> False
         Tip _ _ -> True
-        Bin _ _ l' r' -> noNilInSet l' && noNilInSet r'
+        Bin _ l' r' -> noNilInSet l' && noNilInSet r'
 
--- Invariant: The Mask is a power of 2.  It is the largest bit position at which
---            two elements of the set differ.
-maskPowerOfTwo :: IntSet -> Bool
-maskPowerOfTwo t =
+-- Invariants:
+-- * All keys in a Bin start with the Bin's shared prefix.
+-- * All keys in the Bin's left child have the Prefix's mask bit unset.
+-- * All keys in the Bin's right child have the Prefix's mask bit set.
+prefixOk :: IntSet -> Property
```

Comment:
For test performance reasons, I suspect we want to consider validity of the prefix of a `Bin` relative to that of its parent when making the test recursive, treating the (peculiarly simple) case of the root specially.

Reply:
Fixed 🤦 

> For test performance reasons, I suspect we want to consider validity of the prefix of a Bin relative to that of its parent when making the test recursive, treating the (peculiarly simple) case of the root specially.

Is the trouble really worth it?

```
$ time cabal run intset-properties
...
All 60 tests passed (0.21s)
```

Reply:
For the 0.21s, no. But why does it run for such a short time? Can we make it run more tests? Larger tests? Ideally I want to use more efficient tests and also get that testing time *up*.

Reply:
Separately, I suspect that a relative test is likely to help give more useful information in case of failure, though I don't know that for sure. Regardless, I won't block up this PR for that.

Reply:
> Can we make it run more tests?

This would be the easiest, we can increase the number of quickcheck runs.

> Ideally I want to use more efficient tests and also get that testing time up.

Well, if it helps catch errors. Don't know how one would judge that, except maybe coverage. Otherwise it is a waste.

Reply:
For `Data.Sequence`, that's helpful for dealing with "special casing" in some parts of the code. Maybe it's overkill here.